Extension syntax documentation. (#190)
pals-project/pals: source/miscellaneous.md

(s:extension-syntax)=
## Extension Syntax

An Extension is data added to a PALS lattice that is not part of the PALS standard.
For example, information specific to a particular machine that has no analogue with other
machines can be put in an extension. A list of "registered" extensions is in the 
[Extension List](#c:extension-list) section.

An extension section in a lattice file has a root node and sub-tree below this root
node is part of the extension. Extensions may be marked in a PALS lattice file using
an `extension` node to mark the root node. The syntax is:
```{code} yaml
<name>:
  - extension: <extension-type>
  ... extension info ...
```
where `<name>` is any name (since a given extension type might appear in multiple places in
the lattice file), and `<extension-type>` is the type of the extension. Example:
```{code} yaml
synchrotron-connect:
  extension: Cornell-CESR-Connections
  alarm-system:
    ...
```

Alternatively, Extension root node names may be "registered" using an `extension-names` node 
to be used as the root of the extension sub-tree. Example:
```{code} yaml
PALS:
  extension-names:
    -name: db_node

>     -name: short_name

Reviewer: Would a list of string for `extension-names` be sufficient or do we plan more keys in that list?

Simplification:
```suggestion
    - db_node
    - short_name
```

This includes element `B1a`. 

Note: Pattern matching is not supported in the `value` expression.

%---------------------------------------------------------------------------------------------------
(s:extension-syntax)=
## Extension Syntax

> An Extension is data added to a PALS lattice that is not part of the PALS standard.

Reviewer: Personally, I would probably keep this whole section in `extensions.md`. It is short and to the point, while `Misc` is a catch-all when we run out of section titles :D

Author: Moved.

For example, information specific to a particular machine that has no analogue with other

> machines can be put in an extension. A list of "registered" extensions is in the 

Reviewer 2: ```suggestion
machines can be put in as an extension. A list of "registered" extensions is in the 
```

[Extension List](#c:extension-list) section.

An extension section in a lattice file has a root node and any sub-nodes of this root
node in the information tree is part of the extension. Extensions may be marked in a PALS lattice file using

> an `extension` node to mark the root node. The syntax is:

Reviewer: The phrasing here is quite different than what we used so far in the standard, can can be mistaken for the `PALS root node`. I would suggest instead:

```suggestion
An extension can be introduced at any dictionary entry within the `PALS` information tree.
If the value of that dictionary entry contains an `extension` key, that entry is treated as the start of extension metadata.
The `extension` value identifies the extension type, and the remaining key-value pairs in that dictionary entry belong to the extension rather than to the PALS standard.

The syntax is:
```

Author: major reorg. Please take a look.

```{code} yaml
<name>:
  extension: <extension-type>
  ... extension info ...
```
where `<name>` is any name (since a given extension type might appear in multiple places in
the lattice file), and `<extension-type>` is the type of the extension. Example:
```{code} yaml
synch_connect:                         # Extension name
  extension: Cornell_CESR_Connect      # Extension type

>   alarm-system:                        # Extension stuff ...

Reviewer: PALS naming convention is alphanumerics and underscores:
```suggestion
  alarm_system:                        # Extension stuff ...
```

Author: Changed.

    ...
```

> In this example, the `synch_connect` node is the extension root node. The type of extension is

Reviewer: ```suggestion
In this example, the `synch_connect` node is the extension entry. The type of extension is
```

Author: major reorg. Please take a look.

> `Cornell-CESR-Connect`. Any sub-nodes below the `synch_connect` node will excluded from any

Reviewer 2: ```suggestion
`Cornell-CESR-Connect`. Any sub-nodes below the `synch_connect` node will be excluded from any
```

PALS validation process.

Alternatively, extension root node names may be "registered" using an `extension_labels` node 

> to be used as the root of the extension sub-tree. `extension_labels` has three components:

Reviewer: ```suggestion
Alternatively, extension metadata names may be "registered" using an `extension_labels` attribute 
to be used as the entry of the extension sub-tree. `extension_labels` has three components:
```

Author: major reorg. Please take a look.

```{code} yaml
extension_labels:
  names:            # List of names
  prefixes:         # List of prefixes

>   suffixes:         # List of suffixes

Reviewer: Two things:
A) This names this as `list` but the example below uses a dictionary of entries in each name, prefixes and suffixes.
B) To simplify, we can just show immediately that `extension_labels` shall sit in `PALS:`
```suggestion
PALS:
  extension_labels:
    names:            # Dict of names + description
    prefixes:         # Dict of prefixes + description
    suffixes:         # Dict of suffixes + description
```
We already state this must be in the PALS root node, so this is fine and clear.

Author: Changed.

```

> The `names` section contains a list of extension names. 

Reviewer: Example is not a list:
```suggestion
The `names` section contains a dictionary of extension names, where each key is a name and value is  short description.
```

Author: Please see new wording.

Any dictionary key names in the lattice file that match one of these names is considered an extension.

> Similarly, the `prefixes` and `suffixes` sections contains a list of extension prefixes and suffixes.

Reviewer: ```suggestion
Similarly, the `prefixes` and `suffixes` sections contains a dictionary of extension prefixes and suffixes as keys, and their descriptions as values.
```

Any key names in the lattice file that has a prefix or suffix that matches one of these is considered an extension.
Example:
```{code} yaml
PALS:

>   extension_names:

Reviewer: Just some wordsmithing & avoiding repetition.
- We can clarify we mean dictionary keys.
- We can use either `extensions` (avoid repetition w/ names) or `extra_keys`, `extra_attributes`, `extension_properties`/`extra_properties` to reuse wording we use elsewhere in PALS already.
```suggestion
Any dictionary key names in the lattice file that match one of these names is considered an extension.
Similarly, the `prefixes` and `suffixes` sections contains a list of extension prefixes and suffixes.
Any key names in the lattice file that has a prefix or suffix that matches one of these is considered an extension.
Example:
```{code} yaml
PALS:
  extension_properties:
```

    names:
      short_name
      blueprint
    prefixes:

>       SciBmad_

Author: I have revised the text to hopefully make this clear. Have a look.

>       SciBmad_

Reviewer: Get it now, thank you!

>       SciBmad_

Reviewer: Noted in person today: the current syntax here is a `{"names": "short_name blueprint"}`, which is not what we want.

Let us fix to valid yaml list:
```suggestion
    names:
      - short_name
      - blueprint
    prefixes:
      - SciBmad_
```

or a with a description:

```suggestion
    names:
      short_name: description of short_name usage
      blueprint: what is a blue print
    prefixes:
      SciBmad_: for the SciBmad ecosystem
```

  facility:
    Q1_vary_long_string:
      short_name: Q1       # Extension matching "short_name".
      SciBmad_connect:     # Extension with "SciBmad_" prefix.
        ... SciBmad connection info ...
```
The `extension_labels` node must appear as a child of the `PALS` root node.
The advantage of using `extension_labels` is that single node extensions are possible

> (there does not have to be an `extension` sub-node).

Reviewer: ```suggestion
The advantage of using `extension_labels` is that single key extensions are possible
(there does not have to be an `extension` key).
```
